TypeCobol.Analysis/Graph/ControlFlowGraph.Dominator.cs

```diff
@@ -0,0 +1,127 @@
+                    if (b.PredecessorEdges == null)
+                        continue;
+                    workSet.Full();
+                    foreach (int r in b.PredecessorEdges)
+                    {
+                        BasicBlock<N, D> p = PredecessorEdges[r];
+                        workSet.And(dominators[p.Index]);
+                    }
+                    workSet.Set(n);//reflexivity
+                    if (!workSet.Equals(dominators[n]))
+                    {
+                        dominators[n].Copy(workSet);
+                        change = true;
+                    }
+                }
+            } while (change);
+
+            return new Tuple<List<int>, BitSet[]>(blocks, dominators);
+        }
+
+        /// <summary>
+        /// Compute the immediate dominance relation.
+        /// </summary>
+        /// <param name="blocks">The blocks of the graph on which to work.</param>
+        /// <param name="dominators">The dominator sets</param>
+        /// <param name="duplicate">true to create new sets, rather then pruning dominators set</param>
+        /// <returns>The immediate dominance sets</returns>
+        public BitSet[] ComputeImmediateDominators(List<int> blocks, BitSet[] dominators, bool duplicate = true)
+        {
+            BitSet[] idoms = duplicate ? new BitSet[dominators.Length] : dominators;
+            foreach (var b in blocks)
+            {   
+                if (duplicate)
+                {
+                    idoms[b] = new BitSet(dominators.Length);
+                    idoms[b].Copy(dominators[b]);
+                }
+                idoms[b].Clear(b); //Remove reflexivity
+            }
+            BitSet tmp = new BitSet();
+            foreach (var r in blocks)
+            {
+                if (r == this.RootBlock.Index) continue;
+                tmp.Copy(idoms[r]);
+                foreach (var s in blocks)
+                {
+                    if (tmp.Get(s))
+                    {
+                        foreach (var t in blocks)
+                        {
+                            if (t == s) continue;
+                            if (tmp.Get(t) && idoms[s].Get(t))
+                            {
+                                idoms[r].Clear(t);
+                            }
+                        }
+                    }
+                }
+            }
+            return idoms;
+        }
+
+        /// <summary>
+        /// Dump a set of Dominators
+        /// </summary>
+        /// <param name="dominators">The set of Dominators</param>
+        /// <param name="writer"></param>
+        public void DumpDominators(BitSet[] dominators, System.IO.TextWriter writer)
+        {
+            for (int i = 0; i < dominators.Length; i++)
+            {
+                BitSet doms = dominators[i];
+                if (doms != null)
+                {
+                    BasicBlock<N, D> b = this.AllBlocks[i];
```

Comment:
Is `b.Index` not always equal to `i` ? Intuitively I thought that this method should be static because it does not depend on the current CFG (except if we include other info from the block).
Also why the double braces ? Is it a common notation ?

Reply:
Yes done in 
d9ea371

Reply:
and 
9dcd9e1 double bracket removed